crypto/hpke/hpke_util.c

```diff
@@ -173,3 +411,106 @@ EVP_KDF_CTX *ossl_kdf_ctx_create(const char *kdfname, const char *mdname,
     }
     return kctx;
 }
+
+/*
+ * @brief look for a label into the synonym tables, and return its id
+ * @param st is the string value
+ * @param synp is the synonyms labels array
+ * @param arrsize is the previous array size
+ * @return 0 when not found, else the matching item id.
+ */
+static uint16_t synonyms_name2id(const char *st, const synonymttab_t *synp,
+                                size_t arrsize)
+{
+    size_t i, j;
+
+    for (i = 0; i < arrsize; ++i) {
+        for (j = 0; j < OSSL_NELEM(synp[i].synonyms); ++j) {
+            if (OPENSSL_strcasecmp(st, synp[i].synonyms[j]) == 0)
+                return synp[i].id;
+        }
+    }
+    return 0;
+}
+
+/*
+ * @brief map a string to a HPKE suite based on synonym tables
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ */
+int ossl_hpke_str2suite(const char *suitestr, OSSL_HPKE_SUITE *suite)
+{
+    uint16_t kem = 0, kdf = 0, aead = 0;
+    char *st = NULL, *instrcp = NULL;
+    size_t inplen;
+    int labels = 0, result = 0;
+    int delim_count = 0;
+    const char OSSL_HPKE_STR_DELIMSTR[] = { OSSL_HPKE_STR_DELIMCHAR, '\0' };
+
+    if (suitestr == NULL || suitestr[0] == 0x00 || suite == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    /* somebody might someday extend in an untested way... */
+    if (OPENSSL_strnlen(OSSL_HPKE_STR_DELIMSTR, 1) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+    inplen = OPENSSL_strnlen(suitestr, OSSL_HPKE_MAX_SUITESTR);
+    if (inplen >= OSSL_HPKE_MAX_SUITESTR)
+        return 0;
+    /*
+     * we don't want a delimiter at the end of the string;
+     * strtok() doesn't care about that, so we should
+     */
+    if (suitestr[inplen - 1] == OSSL_HPKE_STR_DELIMCHAR)
+        return 0;
+    /* We want exactly two delimiters in the input string */
+    for (st = (char *)suitestr; *st != '\0'; st++) {
+        if (*st == OSSL_HPKE_STR_DELIMCHAR)
+            delim_count++;
+    }
+    if (delim_count != 2)
+        return 0;
+
+    /* Duplicate `suitestr` to allow its parsing  */
+    instrcp = OPENSSL_memdup(suitestr, inplen + 1);
+    if (instrcp == NULL)
+        goto fail;
+
+    /* See if it contains a mix of our strings and numbers */
+    st = strtok(instrcp, OSSL_HPKE_STR_DELIMSTR);
```

Comment:
strtok isn't thread safe...

Reply:
sure, can change to ``strtok_r()`` - would that do it? 
I played with that last night, so no problem doing that.

Reply:
Unfortunately I'm not sure how portable it is....does Windows have it for example?

Reply:
@mattcaswell s390xcap.c has some strtok's in it also.

Reply:
Yeah. I noticed. And e_afalg.c. Maybe we should have an `OPENSSL_strtok`...(or probably `OPENSSL_strok_r`)

Reply:
just pushed that

Reply:
yeah the other ``strtok()`` calls and windows were why I didn't make that change last night. I did some some talk of a ``strtok_s()`` on windows - I guess we'll see what the windows CI builds make  of the latest push

Reply:
yep - windows didn't like ``strtok_r()`` when I tried it last night https://github.com/sftcd/openssl/actions/runs/3528911443/jobs/5919489757
I'll check some more into ``strtok_s()`` ...

Reply:
The problem is we have to support many weird and wonderful platforms (VMS, Nonstop, etc) - so even if we get something that works on Linux and Windows - it doesn't necessarily work everywhere.

I think there are probably 2 options:
1) Avoid use of strtok altogether (can we rewrite not to use it?)
2) Implement an `OPENSSL_strtok_r` which detects (somehow) whether we have strtok_r and uses it if we have (or a Windows equivalent) and if not falls back to some from-scratch implementation

Possibly (2) could be done in some follow-on PR

Reply:
I'll look into (1) a bit, which may become (2) quite quickly:-)

Reply:
(2) should be done using thread-local storage.
I don't think that Openssl have any code or abstraction layer for this feature.

Reply:
> (2) should be done using thread-local storage.

If we go with (2) then we would implement `strtok_r` which is thread safe anyway - no need for thread-local storage

> I don't think that Openssl have any code or abstraction layer for this feature.

`CRYPTO_THREAD_LOCAL`

Reply:
so the latest commit implements (1) and doesn't seem horrible